torch/csrc/api/src/optim/adamw.cpp

```diff
@@ -0,0 +1,70 @@
+#include <torch/optim/adamw.h>
+
+#include <torch/csrc/autograd/variable.h>
+#include <torch/nn/module.h>
+#include <torch/serialize/archive.h>
+#include <torch/utils.h>
+
+#include <ATen/ATen.h>
+
+#include <cmath>
+#include <functional>
+
+namespace torch {
+namespace optim {
+
+AdamWOptions::AdamWOptions(double learning_rate)
+    : learning_rate_(learning_rate) {}
+
+void AdamW::step() {
+  for (size_t i = 0; i < parameters_.size(); ++i) {
+    Tensor p = parameters_.at(i);
+    if (!p.grad().defined()) {
+      continue;
+    }
+
+
+    auto& exp_average = buffer_at(exp_average_buffers, i);
+    auto& exp_average_sq = buffer_at(exp_average_sq_buffers, i);
+
+    buffer_at(step_buffers, i) += 1;
+
+    exp_average.mul_(options.beta1_).add_(p.grad(), 1 - options.beta1_);
+    exp_average_sq.mul_(options.beta2_)
+        .addcmul_(p.grad(), p.grad(), 1 - options.beta2_);
+
+    Tensor denom = exp_average_sq;
+    if (options.amsgrad_) {
+      auto& max_exp_average_sq = buffer_at(max_exp_average_sq_buffers, i);
+      max_exp_average_sq = torch::max(max_exp_average_sq, exp_average_sq);
+      denom = max_exp_average_sq;
+    }
+
+    const auto bias_correction1 =
+        1 - std::pow(options.beta1_, buffer_at(step_buffers, i));
+    const auto bias_correction2 =
+        1 - std::pow(options.beta2_, buffer_at(step_buffers, i));
+    const auto step_size =
+        options.learning_rate_ * std::sqrt(bias_correction2) / bias_correction1;
+
+    NoGradGuard guard;
+
+    if(options.weight_decay_ > 0) {
+        Tensor decoupled_weight_decay = p * (-options.weight_decay_ * step_size);
```

Comment:
Please note that this implementation doesn't match the original paper, nor the python implementation. 
To match the paper, this should be `-options.weight_decay * options.learning_rate`.

Even then, there would be a difference with the pytorch implementation, because epsilon is added to `sqrt(exp_average_sq)`, while it is added to `sqrt(exp_average_sq / bias_correction2)` in python. This essentially changes the scale of `epsilon` and makes both implementation incompatible hyperparameter-wise

Reply:
Note: I didn't check adabound.

Reply:
See in https://github.com/pytorch/pytorch/pull/22628 for the discussion about `epsilon`.

Reply:
I'll let others who are more involved work on this, then, to make it available.